Tests/ApolloTests/StarWars/API.swift

```diff
@@ -336,65 +406,82 @@ public final class HeroDetailsQuery: GraphQLQuery {
   public struct Data: GraphQLMappable {
     public let hero: Hero?
 
-    public init(reader: GraphQLResultReader) throws {
-      hero = try reader.optionalValue(for: Field(responseName: "hero", arguments: ["episode": reader.variables["episode"]]))
+    public init(values: [Any?]) {
+      hero = values[0] as! Hero?
     }
 
     public struct Hero: GraphQLMappable {
       public let __typename: String
       public let name: String
-
       public let asHuman: AsHuman?
       public let asDroid: AsDroid?
 
-      public init(reader: GraphQLResultReader) throws {
-        __typename = try reader.value(for: Field(responseName: "__typename"))
-        name = try reader.value(for: Field(responseName: "name"))
-
-        asHuman = try AsHuman(reader: reader, ifTypeMatches: __typename)
-        asDroid = try AsDroid(reader: reader, ifTypeMatches: __typename)
+      public init(values: [Any?]) {
+        __typename = values[0] as! String
+        name = values[1] as! String
+        asHuman = values[2] as! AsHuman?
```

Comment:
Yes, fragments (both named once and merged object type specific ones like `asHuman`) are treated as separate types. I started out with a model where these fragments were generated as protocols ('interfaces') and the query-specific selection sets implemented them, but that broke down due to limitations of the Swift type system (properties defined in protocols are invariant).

Swift has enums with associated values as a way to do exhaustive pattern matching, but that isn't really a natural fit for cases like this.